crates/bevy_tasks/src/task.rs

```diff
@@ -36,25 +76,134 @@ impl<T> Task<T> {
+    /// Canceling tasks is unsupported on the web, and this is the same as awaiting the task.
     pub async fn cancel(self) -> Option<T> {
-        self.0.cancel().await
+        cfg::web! {
+            if {
+                // Await the task and handle any panics.
+                match self.0.recv().await {
+                    Ok(Ok(value)) => Some(value),
+                    Err(_) => None,
+                    Ok(Err(panic)) => {
+                        // drop this to prevent the panic payload from resuming the panic on drop.
+                        // this also leaks the box but I'm not sure how to avoid that
+                        core::mem::forget(panic);
+                        None
+                    }
+                }
+            } else {
+                // Wait for the task to become canceled
+                self.0.cancel().await
+            }
+        }
     }
 
     /// Returns `true` if the current task is finished.
     ///
-    ///
     /// Unlike poll, it doesn't resolve the final value, it just checks if the task has finished.
     /// Note that in a multithreaded environment, this task can be finished immediately after calling this function.
     pub fn is_finished(&self) -> bool {
-        self.0.is_finished()
+        cfg::web! {
+            if {
+                // We treat the task as unfinished until the result is sent over the channel.
+                !self.0.is_empty()
+            } else {
+                // Defer to the `async_task` implementation.
+                self.0.is_finished()
+            }
+        }
     }
 }
 
 impl<T> Future for Task<T> {
     type Output = T;
 
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        Pin::new(&mut self.0).poll(cx)
+    cfg::web! {
+        if {
+            fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+                // `recv()` returns a future, so we just poll that and hand the result.
+                let recv = core::pin::pin!(self.0.recv());
+                match recv.poll(cx) {
+                    Poll::Ready(Ok(Ok(value))) => Poll::Ready(value),
+                    // NOTE: Propagating the panic here sorta has parity with the async_executor behavior.
+                    // For those tasks, polling them after a panic returns a `None` which gets `unwrap`ed, so
+                    // using `resume_unwind` here is essentially keeping the same behavior while adding more information.
+                    Poll::Ready(Ok(Err(_panic))) => crate::cfg::switch! {{
+                        crate::cfg::std => {
+                            std::panic::resume_unwind(_panic)
+                        }
+                        _ => {
+                            unreachable!("catching a panic is only possible with std")
+                        }
+                    }},
+                    Poll::Ready(Err(_)) => panic!("Polled a task after it was cancelled"),
```

Comment:
Isn't cancellation impossible in web?
Or I guess this could happen if we are in no-std and the web task panics, which means that the channel gets closed? If that's the case, maybe the panic message could be updated for clarity

Reply:
This code was just moved from `wasm_task.rs`, none of this logic is now. The only time this will error is if we get a `async_channel::RecvError` which means the sender-side is closed and the channel is empty, which means the task has already run and been awaited.

So it's correct to panic here. The message could be better though. It's trying to juggle the `async_tasks` terminology as best it can.

Reply:
Do you have a suggestion for a better error message?

Reply:
Done.